Src/FluentAssertions/Equivalency/AssertionRuleEquivalencyStep.cs

```diff
@@ -83,6 +83,8 @@ private bool ExecuteAssertion(IEquivalencyValidationContext context)
 
             if (subjectIsValidType && expectationIsValidType)
             {
+                AssertionScope.Current.ResolveContext();
```

Comment:
I'm struggling to find out why this is necessary.
I have the branch open and see that two tests fails without it, I just can comprehend _why_?
`AssertionScope` is not very friendly to debug/comprehend for _my_ brain.
(I'm not asking you to solve all our problems)

For `When_property_of_other_is_null_the_failure_message_should_not_complain_about_its_type` when it's changed to pass, `DetermineCallerIdentity` is still called twice 😢.

Reply:
This is because it is deferring the calculation of the context string, and the calculation depends on the current node. Through this path, the current node changes, so if we wait until after that, we're calculating it with a different node. Calling `ResolveContext` here locks in the computed value for the previous current node value.

Reply:
@logiclrd , your explaination indicates that this should be documented. If maintainers like Jonas are not able to understand, you should add comments and maybe more specific method name like `BuildContextString`. "Build" is related to the func of the `Lazy` instance which can be "complex building procedure".

@jnyrup , I'm wondering that only 2 tests are failing. This indicates that only 2 of thousand of tests are using the scenario which is described by Jonathan.

Reply:
I don't think that's _that_ unusual, because the scenario involved requires that you use a pattern of:
```
result.Should().BeEquivalentTo(
  expectation,
  o => o.Using(c => c.Should().BeSomethingElse());
```
With this pattern, the `.Using()` is re-entering the entire assertion stack from the top again. This is where deferring the calculation becomes a problem, because the existing unit tests indicate that a failure within the `.Using()` should be _reported_ as a failure of the outer `result.Should()`, but the deferred calculation makes it find the `c.Should()` instead.

This can only occur when you have this re-entrant structure, and the vast majority of the unit tests do not.

I will document this. :-)

Reply:
What we really want is to start a new nested `AssertionScope` that does not capture the `c.Should()`. Can't we make that more explicit?

Reply:
I'm not entirely certain, to be honest -- my familiar with the codebase only started a few weeks ago :-P

Thinking about it, though, it seems that the key to this comes down to how it locates that context. It's looking through the stack trace for the first non-FluentAssertions frame it finds. This change is needed because when the user writes:
```
result.Should().BeEquivalentTo(
  expectation,
  o => o.Using(c => c.Should().BeSomethingElse());
```
...the stack trace starts in user code, then goes into FluentAssertions for `result.Should().BeEquivalentTo()`, then goes _back_ to the user code for `c => c.Should().BeSomethingElse()`, which then goes back into FluentAssertions.

Off the top of my head, initially I was thinking about perhaps making it so that an `AssertionScope` can be given a stack trace to use for resolving the context, so that the re-entrant call actually uses the outer call's stack trace and doesn't see the `c => c.Should().BeSomethingElse()` call at all. But then another thought occurred to me: What if the stack trace were scanned in the opposite order, so that instead of searching for the first frame that isn't in FluentAssertions, we search for the last frame that _is_, and use the frame below that for resolving the context? That would then _always_ use the outermost active call site into FluentAssertions at any point.

Reply:
Okay, my first stab at this produced tons of test failures, because in the case of `delegate.Should().Throw(..)` we have this exact scenario in reverse.

Reply:
I've pushed a commit that I think does the trick.

The trickiest part was the `async` machinery, which gums up the works with stack frames that aren't in FluentAssertions but do need to be ignored. Because of the fact that `action.Should().Throw(..)` type assertions have this exact same structure, where the stack frames at the bottom of the stack will find `action.Should()` but the message should be finding the subject within the "..", I had to create a mechanism for assertions on delegates to indicate that the `CallerIdentifier` logic should find the second instance of user code calling into FluentAssertions and not the first. This only activates if the `action.Should()` is the very first call into FluentAssertions on the call stack.

Reply:
> What if the stack trace were scanned in the opposite order, so that instead of searching for the first frame that isn't in FluentAssertions, 

That is exactly what I think might fix the problem.